crypto/src/dex/trading_pair.rs

```diff
@@ -121,6 +123,40 @@ impl TryFrom<[u8; 64]> for TradingPair {
     }
 }
 
+/// Represents a trading pair in R1CS.
+pub struct TradingPairVar {
+    pub asset_1: asset::AssetIdVar,
+    pub asset_2: asset::AssetIdVar,
+}
+
+impl AllocVar<TradingPair, Fq> for TradingPairVar {
+    fn new_variable<T: std::borrow::Borrow<TradingPair>>(
+        cs: impl Into<ark_relations::r1cs::Namespace<Fq>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: ark_r1cs_std::prelude::AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        let ns = cs.into();
+        let cs = ns.cs();
+        let trading_pair = f()?.borrow().clone();
+        let asset_1 = AssetIdVar::new_variable(cs.clone(), || Ok(trading_pair.asset_1()), mode)?;
+        let asset_2 = AssetIdVar::new_variable(cs, || Ok(trading_pair.asset_2()), mode)?;
+        // A valid `TradingPair` has a canonical ordering.
+        // Constrain: asset_1 <= asset_2.
+        asset_1
+            .asset_id
+            .enforce_cmp(&asset_2.asset_id, core::cmp::Ordering::Less, true)?;
```

Comment:
I think this doesn't work because `enforce_cmp` [expects](https://docs.rs/ark-r1cs-std/latest/ark_r1cs_std/fields/fp/enum.FpVar.html#method.enforce_cmp) the variables to be `<= (p - 1)/2` in size. The other places where we use the `enforce_cmp` method should be checked too.